### ISSUE TITLE
'contentType is required' error when adding a chat member

### ISSUE THREAD

**Issue by darrenparkinson** [microsoftgraph/msgraph-sdk-go#248]
This looks like it might be the same issue as #205 but I'm getting it when trying to add a member to a chat.

I'm essentially using [the code from here](https://docs.microsoft.com/en-us/graph/api/chat-post-members?view=graph-rest-1.0&tabs=go) and I'm getting the `contentType is required` error, but the member is added successfully.

```go
func (m *MSTeamsModel) AddMember(chatid, email string) error {
	conversationMember := graphmodels.NewConversationMember()
	odatatype := "#microsoft.graph.aadUserConversationMember"
	conversationMember.SetOdataType(&odatatype)
	roles := []string{
		"owner",
	}
	conversationMember.SetRoles(roles)
	additionalData := map[string]interface{}{
		"user@odata.bind": fmt.Sprintf("https://graph.microsoft.com/v1.0/users('%s')", email),
	}
	conversationMember.SetAdditionalData(additionalData)
	visibleHistoryStartDateTime, _ := time.Parse(time.RFC3339, "0001-01-01T00:00:00Z")
	conversationMember.SetVisibleHistoryStartDateTime(&visibleHistoryStartDateTime)
	_, err := m.Client.ChatsById(chatid).Members().Post(conversationMember)
	if err != nil {
		printOdataError(err)
		return fmt.Errorf("error adding member: %w", err)
	}
	return nil
}
```

Versions from go.mod when downloaded originally:

```
github.com/Azure/azure-sdk-for-go/sdk/azcore v1.1.0 // indirect
github.com/Azure/azure-sdk-for-go/sdk/azidentity v1.1.0 // indirect
github.com/Azure/azure-sdk-for-go/sdk/internal v1.0.0 // indirect
github.com/microsoft/kiota-abstractions-go v0.8.2 // indirect
github.com/microsoft/kiota-authentication-azure-go v0.3.1 // indirect
github.com/microsoft/kiota-http-go v0.5.2 // indirect
github.com/microsoft/kiota-serialization-json-go v0.5.5 // indirect
github.com/microsoft/kiota-serialization-text-go v0.4.1 // indirect
github.com/microsoftgraph/msgraph-sdk-go v0.34.0 // indirect
github.com/microsoftgraph/msgraph-sdk-go-core v0.27.0 // indirect
```

Just performed a `go get -u github.com/microsoftgraph/msgraph-sdk-go` which updated these -- issue persists:

```
go: downloading github.com/microsoft/kiota-http-go v0.6.0
go: upgraded github.com/microsoft/kiota-http-go v0.5.2 => v0.6.0
go: upgraded github.com/microsoftgraph/msgraph-sdk-go v0.34.0 => v0.35.0
```

Many thanks for any information you can provide.

**Comment by darrenparkinson** [comment]
Great, thanks. After trying that, I still got the same error.  For reference, the function now looks like this:

```go
func (m *MSTeamsModel) AddMember(chatid, email string) error {
	conversationMember := graphmodels.NewAadUserConversationMember()
	roles := []string{
		"owner",
	}
	conversationMember.SetRoles(roles)
	additionalData := map[string]interface{}{
		"user@odata.bind": fmt.Sprintf("https://graph.microsoft.com/v1.0/users/%s", email),
	}
	conversationMember.SetAdditionalData(additionalData)
	visibleHistoryStartDateTime, _ := time.Parse(time.RFC3339, "0001-01-01T00:00:00Z")
	conversationMember.SetVisibleHistoryStartDateTime(&visibleHistoryStartDateTime)
	_, err := m.Client.ChatsById(chatid).Members().Post(conversationMember)
	if err != nil {
		printOdataError(err)
		return fmt.Errorf("error adding member: %w", err)
	}
	return nil
}
```
Finally, the serialised body is as follows:

```json
{
    "@odata.type": "#microsoft.graph.aadUserConversationMember",
    "user@odata.bind": "https://graph.microsoft.com/v1.0/users/replacedemail@domain.com",
    "roles": [
        "owner"
    ],
    "visibleHistoryStartDateTime": "0001-01-01T00:00:00Z"
}
```
Hope that helps.
Thanks again.

**Comment by baywet** [comment]
Thanks for the additional information. The payload the SDK is sending looks correct (compared with the documentation).
If you take the same exact payload and run it through [Graph explorer](https://aka.ms/ge), do you get a working response?

**Comment by darrenparkinson** [comment]
Yes, that works fine.  See attached image.

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/4262917/186870732-ca5826e0-4427-4665-ac3d-c833ac61cafb.png">

**Comment by baywet** [comment]
Thanks for sharing this additional information. At this point my last suggestion would be to check that your application has the right permissions. From the authentication provider you can get the access token provider. And you can call that to get one of the tokens generated. Feed that to jwt.ms and check whether it has the required scopes indicated by the documentation. Do not share the access token with anyone or here.

**Comment by darrenparkinson** [comment]
Hi, yeah, it all works fine -- I can make other requests and this request does actually work, I just get that weird error back which is invalid since contentType isn't required, but of course it comes back as an `error` which in normal go code should mean it failed (even though it didn't) -- hope that makes sense...

**Comment by baywet** [comment]
The sdk is built in a way that if it gets an error response code from the service (400-500) it will return an error.
But now I'm noticing something else,  the service is documented to return a 201 but the generated code returns something in addition to the error.  That means the sdk is failing trying to parse the response,  but it shouldn't even try. 
I believe there might be two different issues here.  The sdk should simply return nil when a service documents a response but replies with 201. And the metadata probably needs correction as well.

**Comment by baywet** [comment]
Update: put together a PR to address the issue in the request adapter. https://github.com/microsoft/kiota-http-go/pull/23

**Comment by baywet** [comment]
Update, the fix has been released, running `go get -u github.com/microsoft/kiota-http-go` should get you sorted.
As per the metadata definition, it's a "gap in the protocol". Effectively the conversation members are just a collection of members, standard OData conventions dictate the service to either return a 200 with a resource representation or a 201 on POST. But there's not annotation for the service to say "I'm only ever going to return 201", which means the metadata (and the generated SDK) are correct, although not precise enough.
Closing.

**Comment by darrenparkinson** [comment]
Awesome, thanks for that. Great turnaround. I’ll give that a go tomorrow.

**Comment by baywet** [comment]
Right, thanks for the callout and sorry for missing it originally.
Authored https://github.com/microsoft/kiota/pull/1800 which will close that issue once merged. And the results will go in the next weekly generation (next week).

**Comment by darrenparkinson** [comment]
OK, great, thanks for the quick turnaround again.  Appreciate it.  So just to be clear, the type assertion in `SendAsync` should no longer panic once that update is applied?

**Comment by baywet** [comment]
correct. Once the PR is merged and a new generation is kicked, your application will get a nil value, nil error and no panic on an empty response body (201 or otherwise)
